File: Classes/View Controllers/Player/EqualizerPointView.m
//
//  EqualizerPointView.m
//  iSub
//
//  Created by Ben Baron on 11/19/11.
//  Copyright (c) 2011 Ben Baron. All rights reserved.
//

#import "EqualizerPointView.h"
#import "BassParamEqValue.h"

#define myWidth 30
#define myHeight 30

@implementation EqualizerPointView
@synthesize frequency=_frequency, gain=_gain, handle=_handle, eqValue=_eqValue;

- (instancetype)initWithCGPoint:(CGPoint)point parentSize:(CGSize)size {
    if (self = [super initWithFrame:CGRectMake(0, 0, myWidth, myHeight)]) {
		_parentSize = size;
		self.center = point;
		
		_position.x = point.x / size.width;
		_position.y = point.y / size.height;
		_handle = 0;
		
		self.image = [UIImage imageNamed:@"eqView.png"];
		
		self.userInteractionEnabled = YES;
		
		BASS_DX8_PARAMEQ p = BASS_DX8_PARAMEQMake(_frequency, _gain, DEFAULT_BANDWIDTH);
		_eqValue = [[BassParamEqValue alloc] initWithParameters:p];
    }
    return self;
}

- (CGFloat)percentXFromFrequency:(NSUInteger)frequency {
	return (log2(frequency) - 5) / 9;
}

- (CGFloat)percentYFromGain:(CGFloat)gain {
	return .5 - (gain / (CGFloat)(MAX_GAIN * 2));
}

- (instancetype)initWithEqValue:(BassParamEqValue *)value parentSize:(CGSize)size {
    if (self = [super initWithFrame:CGRectMake(0, 0, myWidth, myHeight)]) {
		_parentSize = size;
		
		CGFloat x = size.width * [self percentXFromFrequency:value.parameters.fCenter];
		CGFloat y = size.height * [self percentYFromGain:value.parameters.fGain];
		self.center = CGPointMake(x, y);
		
        _position.x = [self percentXFromFrequency:value.parameters.fCenter];
		_position.y = [self percentYFromGain:value.parameters.fGain];
		
		self.image = [UIImage imageNamed:@"eqView.png"];
		
		self.userInteractionEnabled = YES;
		
		self.eqValue = value;
    }
    return self;
}

- (void)setCenter:(CGPoint)center
{
	[super setCenter:center];
	
	_position.x = self.center.x / self.parentSize.width;
	_position.y = self.center.y / self.parentSize.height;
}

- (NSUInteger)frequency {
	return exp2f((self.position.x * RANGE_OF_EXPONENTS) + 5);
}

- (CGFloat)gain {
	return (.5 - self.position.y) * (CGFloat)(MAX_GAIN * 2);
}

- (HFX)handle {
	return self.eqValue.handle;
}

- (void)setEqValue:(BassParamEqValue *)value {
	_eqValue = value;
}

- (BassParamEqValue *)eqValue {
	_eqValue.gain = self.gain;
	_eqValue.frequency = self.frequency;
	_eqValue.bandwidth = DEFAULT_BANDWIDTH;
	
	return _eqValue;
}

- (NSComparisonResult)compare:(EqualizerPointView *)otherObject  {
	// Return ordered same if now the same class as me
	if(![otherObject isKindOfClass:[EqualizerPointView class]]) return NSOrderedSame;
		
	CGFloat myX = self.frame.origin.x;
	CGFloat otherX = otherObject.frame.origin.x;
	
	if (myX < otherX) return NSOrderedAscending;
	if (myX > otherX) return NSOrderedDescending;
	
	return NSOrderedSame;
}

/*CGFloat percentXFromFrequency(NSUInteger frequency)
{
	return (log2(frequency) - 5) / 9;
}

CGFloat percentYFromGain(CGFloat gain)
{
	return .5 - (gain / (CGFloat)(MAX_GAIN * 2));
}
 
CGPoint CGPointMakeFromEqValues(NSUInteger frequency, CGFloat gain)
{
	return CGPointMake(percentXFromFrequency(frequency), percentYFromGain(gain));
}*/

@end

File: Classes/View Controllers/Player/EQ and Visualizer/EqualizerPointView.swift
//
//  EqualizerPointView.swift
//  iSub
//
//  Created by Benjamin Baron on 2/1/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import UIKit

private let defaultWidth: CGFloat = 30
private let defaultHeight: CGFloat = 30
private let rangeOfExponents: Float = 9
private let maxGain: Float = 6

private let defaultBandwidth: Float = 18

private func percentX(frequency: Float) -> CGFloat {
    return CGFloat((log2(frequency) - 5) / 9)
}

func percentY(gain: Float) -> CGFloat {
    return CGFloat(0.5 - (gain / (maxGain * 2)))
}

final class EqualizerPointView: UIImageView {
    private var _eqValue: BassParamEqValue
    var eqValue: BassParamEqValue {
        get {
            _eqValue.gain = gain
            _eqValue.frequency = frequency
            _eqValue.bandwidth = defaultBandwidth
            return _eqValue
        }
        set {
            _eqValue = newValue
        }
    }
    var position: CGPoint
    var parentSize: CGSize
    var handle: HFX = 0
    
    var frequency: Float { exp2((Float(position.x) * rangeOfExponents) + 5) }
    var gain: Float { (0.5 - Float(position.y)) * Float(maxGain * 2) }
    
    override var center: CGPoint {
        didSet {
            position.x = center.x / parentSize.width
            position.y = center.y / parentSize.height
        }
    }
    
    init(point: CGPoint, parentSize: CGSize) {
        let p = BASS_DX8_PARAMEQ(fCenter: 0, fBandwidth: Float(defaultBandwidth), fGain: 0)
        self._eqValue = BassParamEqValue(parameters: p)
        
        // Scale point to parentSize
        let scaledPoint = CGPoint(x: point.x * parentSize.width, y: point.y * parentSize.height)
        self.position = CGPoint(x: point.x, y: point.y)
        self.parentSize = parentSize
        super.init(frame: CGRect(x: 0, y: 0, width: defaultWidth, height: defaultHeight))
        self.center = scaledPoint  // Set scaled point as center
        self.image = UIImage(named: "eqView")
        self.isUserInteractionEnabled = true
    }
    
    init(eqValue: BassParamEqValue, parentSize: CGSize) {
        print("BEN eqValue.parameters: \(eqValue.parameters)\n")
        self._eqValue = eqValue
        self.parentSize = parentSize
        let x = percentX(frequency: eqValue.parameters.fCenter)
        let y = percentY(gain: eqValue.parameters.fGain)
        self.position = CGPoint(x: x, y: y)
        super.init(frame: CGRect(x: 0, y: 0, width: defaultWidth, height: defaultHeight))
        self.center = CGPoint(x: parentSize.width * position.x, y: parentSize.height * position.y)
        self.image = UIImage(named: "eqView")
        self.isUserInteractionEnabled = true
    }
    
    required init?(coder: NSCoder) {
        fatalError("unimplemented")
    }
    
    func compare(other: EqualizerPointView) -> ComparisonResult {
        let myX = frame.origin.x
        let otherX = other.frame.origin.x
        if myX < otherX {
            return .orderedAscending
        } else if myX > otherX {
            return .orderedDescending
        } else {
            return .orderedSame
        }
    }
}
